distributed_shampoo/utils/tests/shampoo_preconditioner_list_test.py

```diff
@@ -455,6 +459,166 @@ def test_amortized_computation_internal_failure(self) -> None:
+                        )
+                    # Check that warnings are logged for four failed amortized computations.
+                    # The fifth one doesn't raise an exception, so no warning is logged.
+                    self.assertCountEqual(
+                        # Only extracts the first sentence in the warning message for simple comparison.
+                        [r.msg.split(". ", maxsplit=1)[0] for r in cm.records],
+                        [
+                            "Matrix computation failed for factor matrix 0.block_0.0 with exception=ValueError()",
+                            "Matrix computation failed for factor matrix 1.block_0.0 with exception=ValueError()",
+                            "Matrix computation failed for factor matrix 1.block_0.1 with exception=ValueError()",
+                            "Matrix computation failed for factor matrix 1.block_1.0 with exception=ValueError()",
+                        ],
+                    )
+                    self.assertEqual(
+                        mock_amortized_computation.call_count,
+                        self.NUM_AMORTIZED_COMPUTATION_CALLS * (step - 1),
+                    )
+                    step += 1
+
+                    # Case 2: amortized computation fails exactly as often as tolerance (3) -> no error.
+                    for _ in range(2):
+                        with self.assertLogs(level="WARNING") as cm:
+                            self._preconditioner_list.update_preconditioners(
+                                masked_grad_list=masked_grad_list,
+                                step=torch.tensor(step),
+                                perform_amortized_computation=True,
+                            )
+                        # Check that warnings are logged for all failed amortized computations.
+                        self.assertCountEqual(
+                            # Only extracts the first sentence in the warning message for simple comparison.
+                            [r.msg.split(". ", maxsplit=1)[0] for r in cm.records],
+                            [
+                                "Matrix computation failed for factor matrix 0.block_0.0 with exception=ValueError()",
+                                "Matrix computation failed for factor matrix 1.block_0.0 with exception=ValueError()",
+                                "Matrix computation failed for factor matrix 1.block_0.1 with exception=ValueError()",
+                                "Matrix computation failed for factor matrix 1.block_1.0 with exception=ValueError()",
+                                "Matrix computation failed for factor matrix 1.block_1.1 with exception=ValueError()",
+                            ],
+                        )
+                        self.assertEqual(
+                            mock_amortized_computation.call_count,
+                            self.NUM_AMORTIZED_COMPUTATION_CALLS * (step - 1),
```

Comment:
I chose to use a persistent `step` variable for all cases to fully simulate the different cases appearing in succession in one training run. The idea is that this also tests that the `step` argument to `self._preconditioner_list.update_preconditioners` does not influence the intended behaviour when all cases appear in this order in a single run, which is what the current test simulates.

Do you prefer resetting `step` in-between cases? While it might work it seems a bit misleading because the current design of the cases (partially) assumes that there is persistent state between the cases (i.e. that all of them are appearing sequentially in a single run).